### ISSUE TITLE
[Epic #306] Phase 4: reference-game training run to published-comparable score (operator-gated)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#329]
**Epic**: #306 — Atari (ALE) environment + CNN policy
**Phase**: 4 of 4
**Phase Goal**: Prove the stack end-to-end: train on the reference game to a published-comparable score.

## Description

Train DQN and/or PPO on the reference game chosen by the spike (#324) using the AtariEnv (#325/#326) and CNN policy (#327):

- Target: published-comparable score (if Pong: a positive score is decisive).
- Document the compute budget honestly — this is a long GPU run on alc-2, **operator-gated like #134**: the run itself needs operator hardware access and should not be auto-claimed by a builder without an operator arranging the run. The reproducible command, config, and learning-curve report are the reviewable artifacts.
- Report protocol: same as #298 (documented, reproducible command + honest learning-curve report), satisfying epic acceptance criterion #2.

## Acceptance Criteria

- [ ] Documented, reproducible training command (config committed)
- [ ] Learning-curve report with the honest budget (wall-clock, hardware, steps)
- [ ] Published-comparable score reached on the reference game (or an honest analysis of the gap)

## Dependencies

Part of Epic #306. Blocked by #325, #326, #327 (and benefits from #328's backend verdict for choosing the training backend).

---
*Decomposed from Epic #306*

**Comment by rjwalters** [comment]
**Phase 1 spike confirmation (docs/ALE_BINDING_STRATEGY.md, #324)**

Reference game confirmed: **Pong** (small action space, dense ±1 signal, decisive success criterion — any positive score beats the ~−21 random floor; cited baselines DQN ≈18.9 / PPO ≈20.7). Runbook addition from the Option D binding: the alc-2 host must have \`pip install ale-py\` (+ ROM acquisition via \`AutoROM --accept-license\` or the bundled ale-py ROMs) as a prerequisite, since the env runs as a subprocess. Operator-gated long GPU run as originally scoped.

**Comment by rjwalters** [comment]
Curating this issue per user request (explicit instruction, Phase 4 of Epic #306 with full Atari/CNN stack confirmed on main).

## Curator Enhancement

### Context

The full stack required for this run is confirmed on `origin/main` as of the latest merge:

- `src/env/games/atari/` — `AtariEnv` subprocess adapter + `AtariPreprocess`/`PreprocessConfig` (Machado defaults: frame-skip 4, frame-stack 4, sticky-action p=0.25, screen 84x84, life-loss termination)
- `src/policy/atari_cnn.rs` — `NatureDqnQNetwork<B>` + `NatureDqnConfig` (Nature-DQN conv trunk: 3 conv layers -> 3136-wide flat -> 512 fc_common -> q_head; `copy_params_from` for target-net sync)
- `src/train/dqn/` — `DQNTrainerBurn<B,Q>` + `DQNConfig` (Double-DQN, Smooth-L1 loss, hard or soft target sync)
- `envs/atari/ale_worker.py` — ALE subprocess worker (Option D; invoked via `ATARI_PYTHON` env var, default `python3`)
- `src/utils/cuda.rs` — `default_burn_device::<B>()` for backend-agnostic device construction

**alc-2 confirmed ready** (operator): RTX 4090 24 GB, CUDA 12.0, fresh clone, ale-py 0.12 importable by `/usr/bin/python3`.

**Reference game confirmed** (Phase 1 spike, `docs/ALE_BINDING_STRATEGY.md`): **Pong**. Success criterion: any positive mean episode score beats the ~-21 random floor. Published baselines: DQN ~18.9 (Mnih 2015, 50M frames) / PPO ~20.7. Pong typically crosses zero well before 10M frames, making it tractable on a single operator-gated run.

---

### PPO vs DQN scope decision

**One binary, DQN first.** DQN is the Nature-2015 reference baseline and is already fully wired (DQNTrainerBurn + NatureDqnQNetwork). PPO on Atari is feasible (NatureDqnBurnPolicy exists and PPOTrainerBurn is in-tree) but adds scope and a second example binary. The issue's success criterion ("published-comparable score") is cleanest with DQN, which produced the cited baselines. A PPO variant is worth a follow-up issue if DQN's Pong learning curve is positive.

---

### Example binary: `examples/games/atari/train_pong_dqn.rs`

**Layout convention** (inspected from `Cargo.toml` and existing examples): each example lives at `examples/games/<game>/<name>.rs` with a `[[example]]` entry in `Cargo.toml` listing `required-features`. The `atari/` subdirectory does not yet exist under `examples/games/`; create it alongside the binary.

**Cargo.toml entry to add** (after the existing pong examples):

```toml
# Nature-DQN on ALE Pong via the AtariPreprocess subprocess adapter
# (Epic #306, Phase 4 — issue #329). Requires env-atari + cuda (or wgpu for
# non-NVIDIA) at run time; NdArray smoke run works CPU-only.
[[example]]
name = "train_pong_dqn"
path = "examples/games/atari/train_pong_dqn.rs"
required-features = ["training", "env-atari"]
```

Note: `env-atari` is sufficient as a compile guard. The `cuda` feature is a *run-time* backend selection applied via the backend type alias and the compile-time feature flag — see the `[cfg(feature = "cuda")]` / `[cfg(not(feature = "cuda"))]` type alias pattern described in `docs/BURN_BACKENDS.md`. The example must use `#[cfg]` to select the backend type, matching how existing examples parameterize the `B` type alias.

**Backend type alias pattern** (match `train_cartpole_modern.rs` / `docs/BURN_BACKENDS.md`):

```rust
#[cfg(feature = "cuda")]
type B = Autodiff<burn::backend::Cuda>;
#[cfg(feature = "wgpu")]
#[cfg(not(feature = "cuda"))]
type B = Autodiff<burn::backend::Wgpu>;
#[cfg(not(any(feature = "cuda", feature = "wgpu")))]
type B = Autodiff<NdArray<f32>>;
```

Use `thrust_rl::utils::cuda::default_burn_device::<B>()` for device construction.

---

### Wiring: `AtariPreprocess` + `NatureDqnQNetwork` + `DQNTrainerBurn`

**Environment construction** (Machado defaults, no config override needed):

```rust
use thrust_rl::env::games::atari::{AtariPreprocess, PreprocessConfig};

let mut env = AtariPreprocess::new("pong", SEED)?;
// observation_space().shape == [4, 84, 84]; obs_len = 4*84*84 = 28_224
// action_space() == Discrete(6)  (PONG_ACTION_COUNT)
```

`AtariPreprocess::new` uses `PreprocessConfig::default()` (frame-skip 4, frame-stack 4, sticky-action p=0.25, screen 84, life-loss termination true). Pass `ATARI_PYTHON=/usr/bin/python3` in the environment; `AtariEnv` reads `ENV_PYTHON = "ATARI_PYTHON"` and falls back to `"python3"`.

**Observation flattening**: `get_observation()` returns a `Vec<f32>` of length 28,224 (already flattened by the preprocessor: `4*84*84` values in 0.0–1.0). The `DQNTrainerBurn` replay buffer takes flat observations directly — no extra reshape on the Rust side. The reshape to `[B, 4, 84, 84]` happens inside the forward closures passed to `select_action` and `train_step`.

**Q-network + target net construction**:

```rust
use thrust_rl::policy::atari_cnn::{NatureDqnQNetwork, NatureDqnConfig};

let n_actions = 6_i64; // PONG_ACTION_COUNT
let cfg = NatureDqnConfig::default().with_seed(SEED);
let online = NatureDqnQNetwork::<B>::with_config(n_actions as usize, cfg, &device);
// NatureDqnQNetwork has copy_params_from — DQNTrainerBurn accepts a second Q for target.
// DQNTrainerBurn's hard-sync path calls maybe_sync_target with a clone closure;
// the builder SHOULD use the hard-copy path (target_update_interval, not soft tau)
// to match the Mnih 2015 spec.
```

The `DQNTrainerBurn` forward closures must reshape the flat buffer tensors:

```rust
let forward_fn = |q: &NatureDqnQNetwork<B>, o_flat: Tensor<B, 2>| {
    let b = o_flat.dims()[0];
    q.forward(o_flat.reshape([b, 4, 84, 84]))   // -> Q-values [B, 6]
};
```

Pass the same closure type for both the online and target forward calls in `train_step`.

---

### Replay buffer sizing — honest memory math

The in-tree `ReplayBuffer` stores observations as `Vec<f32>` (`obs` and `next_obs` both, 4 bytes each):

```
1 frame slot = 2 * 28,224 * 4 bytes = 225,792 bytes  ≈ 221 KiB
1M-frame buffer                      ≈ 210 GiB   (does not fit in 24 GB VRAM or typical 32 GB host RAM)
```

**Prescribe: `buffer_capacity = 100_000`** (≈ 21 GiB obs+next_obs — also too large for VRAM, lives in host RAM; host RAM on alc-2 must be confirmed by operator but 32 GB is typical). If host RAM is tight, fall back to `50_000` (≈ 10.5 GiB). Document this in the example's module comment with the exact math:

```
buffer at 100k: 2 * 28224 * 4 * 100_000 = ~21.0 GiB (host RAM)
buffer at  50k: 2 * 28224 * 4 *  50_000 = ~10.5 GiB (host RAM)
```

Note for future improvement: if buffer footprint becomes the binding constraint, a u8 frame store (before the 1/255 scale) would reduce obs storage by 4x (~5.3 GiB at 100k). The current `ReplayBuffer` stores `f32` only; a u8 path would require a new buffer type (out of scope for this issue).

---

### Hyperparameters (Mnih 2015 adapted for 5–10M frame budget)

Nature DQN paper values and honest per-budget adaptations:

| Parameter | Mnih 2015 (50M) | This run (~5–10M) |
|---|---|---|
| `learning_rate` | 2.5e-4 (RMSProp) | 2.5e-4 (Adam — use Adam per in-tree precedent) |
| `batch_size` | 32 | 32 |
| `buffer_capacity` | 1M | 100_000 (f32 budget; see RAM math above) |
| `min_buffer_size` | 50_000 | 10_000 |
| `target_update_interval` | 10_000 | 10_000 (hard copy) |
| `gamma` | 0.99 | 0.99 |
| `epsilon_start` | 1.0 | 1.0 |
| `epsilon_end` | 0.1 | 0.1 |
| `epsilon_decay_steps` | 1_000_000 | 1_000_000 (first 1M of 5–10M budget) |
| `max_grad_norm` | 10.0 (clip) | 10.0 |
| `soft_update_tau` | None (hard sync) | None |

**Frame budget**: prescribe `TOTAL_TIMESTEPS` env default of `5_000_000`. Override via `TOTAL_TIMESTEPS` env var (same pattern as `train_cartpole_dqn.rs`). Pong typically goes positive within the first 1–3M frames with Mnih-scale hyperparameters; 5M is a conservative floor for first positive, 10M gives a comfortable margin.

Honest citation: "DQN reaches ~18.9 at 50M frames (Mnih 2015); Pong typically crosses zero well within 10M frames per the spike analysis in `docs/ALE_BINDING_STRATEGY.md`. This run targets ≤10M frames to produce a positive score rather than a ceiling score."

---

### Checkpointing + eval cadence

The in-tree DQN trainer does not have built-in checkpoint saving (weights to disk). For this run, prescribe:
- Log every `LOG_INTERVAL = 10_000` env steps to stdout (same pattern as `train_cartpole_dqn.rs`).
- Emit a `CURVE_CSV=<path>` learning-curve CSV with columns `env_steps,mean_episode_reward` (same schema as the CartPole DQN, A2C, PPO curves).
- For checkpointing: use Burn's `Module::save_file` (or `into_record` + `serde_json`) at a configurable `CHECKPOINT_INTERVAL` (e.g. 500_000 steps) to `checkpoints/pong_dqn_<step>.json`. This is optional-but-recommended for a long run; document the env var to enable it.

---

### Runbook — exact alc-2 commands

The operator has confirmed alc-2 is ready (RTX 4090 24 GB, CUDA 12.0, ale-py 0.12 importable by `/usr/bin/python3`).

**Phase 0 — verify pre-flight (once, before the run)**
```bash
# Confirm ale-py importable
/usr/bin/python3 -c "import ale_py; print(ale_py.__version__)"
# Confirm Pong ROM available
/usr/bin/python3 -c "import ale_py; ale = ale_py.ALEInterface(); ale.loadROM(ale_py.roms.Pong)"
# Confirm CUDA visible
nvidia-smi
```

**Phase 1 — start the training run (in a tmux session)**
```bash
# On alc-2, from the repo root
tmux new-session -s pong_dqn

# Set the Python override (ale-py is at /usr/bin/python3, not the default python3)
export ATARI_PYTHON=/usr/bin/python3

# Optional: enable learning-curve CSV output
export CURVE_CSV="$HOME/pong_dqn_curve.csv"

# Optional: enable checkpointing every 500k steps
export CHECKPOINT_INTERVAL=500000
export CHECKPOINT_DIR="$HOME/pong_dqn_checkpoints"

# Run the training binary (cuda feature selects the RTX 4090)
cargo run --release \
  --features "training,env-atari,cuda" \
  --example train_pong_dqn \
  2>&1 | tee "$HOME/pong_dqn_run.log"
```

PATH note: if `cargo` is not on the default PATH in a fresh tmux (common on alc-* nodes), source the shell profile first: `source ~/.bashrc` or `source ~/.cargo/env`. The `envs/` directory (containing `ale_worker.py`) must exist at the repo root — the default `ATARI_WORKER_SCRIPT` is a relative path (`envs/atari/ale_worker.py`), so run from the repo root, or set `ATARI_WORKER_SCRIPT` to the absolute path.

**Expected wall-clock estimate**: 5M env frames at ~300–600 frame/s throughput on RTX 4090 with frame-skip 4 (i.e. ~75–150 ALE steps/s overhead) is approximately 2–5 hours. The IPC overhead of the subprocess adapter is amortized by frameskip; actual throughput should be validated by the operator in the first few minutes of the run (log output shows steps/s).

**Where results land**:
- Stdout / tmux pane (progress every 10k steps)
- `$HOME/pong_dqn_run.log` — full log
- `$HOME/pong_dqn_curve.csv` — learning-curve CSV
- `$HOME/pong_dqn_checkpoints/` — weight snapshots (if `CHECKPOINT_DIR` is set)

**Phase 2 — commit results**
After the run, commit:
- `docs/research/2026-07-pong-dqn-learning-curve.md` — honest report: hardware, wall-clock, env steps, final mean score, learning-curve discussion, gap analysis vs Mnih 2015
- `docs/research/data/2026-07-pong_dqn_curve.csv` — raw learning-curve CSV

Per the `#298` / epic `#306` report protocol: if Pong goes positive, report the step at which it crossed zero and the final score. If not, report an honest gap analysis (current score, convergence shape, likely blockers).

---

### Scope guard

**No `src/` changes** are required for this phase. The entire stack is already on `main`. The only deliverables are:
1. `examples/games/atari/train_pong_dqn.rs` — new example binary
2. `Cargo.toml` — one new `[[example]]` entry
3. `docs/research/2026-07-pong-dqn-learning-curve.md` — post-run operator report (NOT a builder artifact)
4. `docs/research/data/2026-07-pong_dqn_curve.csv` — raw CSV (post-run)

If during implementation a small genuinely-missing utility is needed (e.g. a per-step frame/s counter helper), justify it in the PR and keep it under 30 LOC. No hyperparameter search, no PPO variant, no src/env or src/policy changes.

---

## Acceptance Criteria (refined split)

### Artifact criteria (CI-checkable — builder deliverable)

- [ ] `examples/games/atari/train_pong_dqn.rs` committed and compiles: `cargo check --features "training,env-atari"` passes on origin/main
- [ ] `[[example]]` entry added to `Cargo.toml` with `required-features = ["training", "env-atari"]`
- [ ] Smoke run passes CPU-side: `TOTAL_TIMESTEPS=200 cargo run --release --features "training,env-atari" --example train_pong_dqn` completes without error (requires ale-py locally; CI skips if not present, matching the existing `env-atari` advisory job pattern)
- [ ] Runbook is committed (either inline in the example's module doc or as `docs/PONG_DQN_RUNBOOK.md`) with exact alc-2 commands and memory math
- [ ] `CURVE_CSV` and `CHECKPOINT_INTERVAL` env vars documented in the example's module doc

### Run criteria (operator-executed on alc-2 — post-merge)

- [ ] Operator runs the committed binary on alc-2 with `--features "training,env-atari,cuda"` per the runbook
- [ ] Run completes (or is cleanly interrupted at a checkpoint) within the authorized GPU budget
- [ ] Mean episode score reaches > 0 (positive Pong), OR a documented honest gap analysis is produced explaining why not
- [ ] Learning-curve CSV committed to `docs/research/data/2026-07-pong_dqn_curve.csv`
- [ ] Learning-curve report committed to `docs/research/2026-07-pong-dqn-learning-curve.md` per the epic #306 / #298 protocol: hardware, wall-clock, steps, score, gap analysis if not at target

---

## Affected Files

Files the builder creates or modifies (all verified to be new or already on `origin/main`):

- `examples/games/atari/train_pong_dqn.rs` — NEW; the DQN training binary (does not yet exist on `origin/main`)
- `Cargo.toml` — EXISTS on `origin/main`; builder adds one `[[example]]` entry

Files created post-run by the operator (not builder artifacts):

- `docs/research/2026-07-pong-dqn-learning-curve.md` — NEW (post-run report)
- `docs/research/data/2026-07-pong_dqn_curve.csv` — NEW (post-run CSV)

Read-only library files (already on `origin/main`, no changes needed):

- `src/env/games/atari/env.rs`, `preprocess.rs`, `mod.rs`, `protocol.rs`
- `src/policy/atari_cnn.rs`
- `src/train/dqn/trainer.rs`, `config.rs`
- `src/utils/cuda.rs`
- `envs/atari/ale_worker.py`

---

## Test Plan

- [ ] Manual smoke run: `TOTAL_TIMESTEPS=200 ATARI_PYTHON=/usr/bin/python3 cargo run --release --features "training,env-atari" --example train_pong_dqn` — verifies the wiring compiles and the subprocess adapter initializes
- [ ] Automated: `cargo check --features "training,env-atari"` in CI (no ale-py required for compilation)
- [ ] Automated: existing atari integration CI job (`atari_integration`, advisory) covers env + preprocess round-trips
- [ ] Learning-curve CSV: set `CURVE_CSV=/tmp/test.csv`, run 200 steps, verify file has header row + at least one data row
- [ ] Operator run on alc-2: full budget, verify positive Pong score or honest gap analysis in the committed report
- [ ] Edge case: run without `ATARI_PYTHON` set — binary should fall back to `python3` and produce an actionable error if ale-py is absent
